src/Backups/BackupEntriesCollector.cpp

```diff
@@ -239,6 +249,7 @@ bool BackupEntriesCollector::tryGatherMetadataAndCompareWithPrevious(std::option
         table_infos.clear();
         gatherDatabasesMetadata();
         gatherTablesMetadata();
+        lockTablesForReading();
```

Comment:
I've written a [test](https://github.com/ClickHouse/ClickHouse/pull/39455/files#diff-b0c6059a60439150d892d295856de7dcd042339e92efbf8df8276419f6791f1dR175) but it didn't help me catch it :(

Reply:
It's because test uses `Replicated` database engine, but concurrent DROP is dangerous only for `Ordinary`.

Reply:
I've added another test, let's check if it will fail
https://github.com/ClickHouse/ClickHouse/pull/39595

Reply:
As other stress tests for race conditions, it may fail like once per 100 or even 1000 runs. 
Also consider adding a functional test, because we run them in CI with Ordinary engine as well (mostly as part of Stress Tests, but anyway)